integration/proxy_tunnel_strategy_test.go

```diff
@@ -0,0 +1,422 @@
+	// a reverse tunnel to the first proxy.
+	p.makeProxy(t)
+	require.Len(t, p.proxies, 2)
+
+	// make sure node doesn't open any reverse tunnel to the second proxy.
+	waitForMaxActiveTunnelConnections(t, p.proxies[1].Tunnel, p.cluster, 0)
+
+	// make sure both proxies are connected to each other.
+	waitForActivePeerProxyConnections(t, p.proxies[0].Tunnel, 1)
+	waitForActivePeerProxyConnections(t, p.proxies[1].Tunnel, 1)
+
+	// make sure we can connect to the node going though any proxy.
+	p.dialNode(t)
+
+	// make sure we can connect to the database going though any proxy.
+	p.dialDatabase(t)
+}
+
+// dialNode starts a client conn to a node reachable through a specific proxy.
+func (p *proxyTunnelStrategy) dialNode(t *testing.T) {
+	for _, proxy := range p.proxies {
+		ident, err := p.node.Process.GetIdentity(types.RoleNode)
+		require.NoError(t, err)
+		nodeuuid, err := ident.ID.HostID()
+		require.NoError(t, err)
+
+		creds, err := GenerateUserCreds(UserCredsRequest{
+			Process:  p.auth.Process,
+			Username: p.username,
+		})
+		require.NoError(t, err)
+
+		client, err := proxy.NewClientWithCreds(
+			ClientConfig{
+				Cluster: p.cluster,
+				Host:    nodeuuid,
+			},
+			*creds,
+		)
+		require.NoError(t, err)
+
+		output := &bytes.Buffer{}
+		client.Stdout = output
+
+		cmd := []string{"echo", "hello world"}
+		err = client.SSH(context.Background(), cmd, false)
+		require.NoError(t, err)
+		require.Equal(t, "hello world\n", output.String())
+	}
+}
+
+func (p *proxyTunnelStrategy) dialDatabase(t *testing.T) {
+	for i, proxy := range p.proxies {
+		connClient, err := postgres.MakeTestClient(context.Background(), common.TestClientConfig{
+			AuthClient: p.dbAuthClient,
+			AuthServer: p.auth.Process.GetAuthServer(),
+			Address:    proxy.GetWebAddr(),
+			Cluster:    p.cluster,
+			Username:   p.username,
+			RouteToDatabase: tlsca.RouteToDatabase{
+				ServiceName: p.cluster + "-postgres",
+				Protocol:    defaults.ProtocolPostgres,
+				Username:    "postgres",
+				Database:    "test",
+			},
+		})
+		require.NoError(t, err)
+
+		result, err := connClient.Exec(context.Background(), "select 1").ReadAll()
+		require.NoError(t, err)
+		require.Equal(t, []*pgconn.Result{postgres.TestQueryResponse}, result)
+		require.Equal(t, uint32(i+1), p.postgresDB.QueryCount())
+
+		err = connClient.Close(context.Background())
+		require.NoError(t, err)
+	}
+}
+
+// makeLoadBalancer bootsraps a new load balancer for proxy instances.
+func (p *proxyTunnelStrategy) makeLoadBalancer(t *testing.T) {
+	lbAddr := utils.MustParseAddr(net.JoinHostPort(Loopback, strconv.Itoa(ports.PopInt())))
+	lb, err := utils.NewLoadBalancer(context.Background(), *lbAddr)
+	require.NoError(t, err)
+
+	t.Cleanup(func() {
+		require.NoError(t, lb.Close())
+	})
+
+	require.NoError(t, lb.Listen())
+	go lb.Serve()
+
+	if p.lb == nil {
+		p.lb = lb
+	} else {
+		t.Error("load balancer already initialized")
+	}
```

Comment:
Can we check for `p.lb != nil` at the beginning of the function and change from `t.Error` to `require.Fail`?